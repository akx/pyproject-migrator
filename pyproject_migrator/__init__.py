import logging

log = logging.getLogger("pyproject-migrator")
