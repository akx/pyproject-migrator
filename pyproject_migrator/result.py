import dataclasses
from pathlib import Path

from pyproject_migrator import log


@dataclasses.dataclass
class Result:
    dest: dict = dataclasses.field(default_factory=dict)
    warnings: list[str] = dataclasses.field(default_factory=list)

    def warn(self, msg: str, path: Path) -> None:
        msg = f"{path}: {msg}"
        self.warnings.append(msg)
        log.warning(msg)

    def assign(self, key: str, value: object) -> None:
        dest, final_key = self._find_dest(key)
        if final_key in dest:
            raise ValueError(f"Key {key} already exists")
        dest[final_key] = value

    def append(self, key: str, value: object) -> None:
        dest, final_key = self._find_dest(key)
        dest.setdefault(final_key, []).append(value)

    def merge(self, t_key: str, vals: dict, *, path: Path) -> None:
        dest, final_key = self._find_dest(t_key)
        dest = dest.setdefault(final_key, {})
        for key, value in vals.items():
            if isinstance(value, list):
                if key in dest:
                    self.warn(
                        f"Appending values in {t_key}->{key}; make sure the order is correct",
                        path,
                    )
                dest.setdefault(key, []).extend(value)
            elif isinstance(value, dict):
                dest.setdefault(key, {}).update(value)
            else:
                if key in dest:
                    self.warn(
                        f"Overwriting {t_key}->{key} (previously {dest[key]!r}) with {value!r}",
                        path,
                    )
                dest[key] = value

    def _find_dest(self, key: str) -> tuple[dict, str]:
        dest = self.dest
        bits = key.split(".")
        final_key = bits.pop()
        for bit in bits:
            dest = dest.setdefault(bit, {})
        return dest, final_key
