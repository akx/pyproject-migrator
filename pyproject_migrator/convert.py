import configparser
import re
from pathlib import Path

from pyproject_migrator.result import Result

PYTEST_WS_SPLIT_KEYS = {"addopts", "norecursedirs", "doctest_optionflags"}

CONFIG_FILES = {
    "mypy.ini",
    "pytest.ini",
    "setup.cfg",
}


def massage_value(value, key, *, ws_split_keys):
    if key in ws_split_keys:
        return [v for v in value.split() if v]
    if isinstance(value, str):
        if "\n" in value:
            return [v for v in value.splitlines() if v]
        if value.lower() == "true":
            return True
        if value.lower() == "false":
            return False
        if value.isnumeric():
            return int(value)
    return value


def translate_config(config: dict, *, ws_split_keys=()) -> dict:
    return {
        key: massage_value(value, key, ws_split_keys=ws_split_keys)
        for key, value in config.items()
    }


FLAKE8_SECTION_RE = re.compile(r"flake8.*|pep8|py(code|doc)style|pep257")
TOX_SECTION_RE = re.compile(r"testenv.*|tox.*")
SETUPTOOLS_SECTION_RE = re.compile(
    r"[bs]dist.*|"
    r"aliases|"
    r"build_ext|"
    r"egg_info|"
    r"entry_points|"
    r"extras|"
    r"files|"
    r"install|"
    r"metadata|"
    r"options.*|"
    r"paths|"
    r"upload_docs|"
    r"wheel"
)


def process_config_file(res: Result, pth: Path):
    config = configparser.ConfigParser()
    config.read(pth)
    setuptools_sections = {}
    flake8_sections = {}
    tox_sections = {}

    for section in config.sections():
        if section == "mypy":
            res.assign("tool.mypy", translate_config(config[section]))
            continue

        if section.startswith("mypy-"):
            res.append(
                "tool.mypy.overrides",
                {
                    "module": section[5:],
                    **translate_config(config[section]),
                },
            )
            continue

        if section.endswith("pytest"):
            translated = translate_config(
                config[section],
                ws_split_keys=PYTEST_WS_SPLIT_KEYS,
            )
            res.merge("tool.pytest.ini_options", translated, path=pth)
            continue
        if section.endswith("isort"):
            translated = translate_config(
                config[section],
                ws_split_keys={"skip"},
            )
            res.assign("tool.isort", translated)
            continue

        if section.startswith("coverage"):
            res.assign(
                f"tool.{section.replace(':', '.')}", translate_config(config[section])
            )
            continue

        if section.startswith("pylint"):
            res.assign(
                f"tool.{section.replace(':', '.')}", translate_config(config[section])
            )
            continue

        if section == "codespell":
            res.assign("tool.codespell", translate_config(config[section]))
            continue

        if section == "build_sphinx":
            res.warn(
                "The build_sphinx section is not supported, "
                "and support will be removed with Sphinx 7 anyway. "
                "See https://www.sphinx-doc.org/en/master/usage/advanced/setuptools.html",
                pth,
            )
            continue

        if TOX_SECTION_RE.match(section):
            tox_sections[section] = config[section]
            continue

        if FLAKE8_SECTION_RE.match(section):
            flake8_sections[section] = config[section]
            continue

        if SETUPTOOLS_SECTION_RE.match(section):
            setuptools_sections[section] = config[section]
            continue

        res.warn(f"unknown section {section}", path=pth)

    if setuptools_sections:
        # TODO: we could allow setting a target toolset and port these.
        res.warn(
            f"setuptools sections {sorted(setuptools_sections.keys())}; porting depends on which toolset to use. "
            "For example, hatch (https://pypi.org/project/hatch/) can port metadata to pyproject.toml.",
            path=pth,
        )

    if flake8_sections:
        res.warn(
            f"flake8-related sections {sorted(flake8_sections.keys())}; porting depends on which toolset to use. "
            "If you're transitioning to ruff, see https://pypi.org/project/flake8-to-ruff/",
            path=pth,
        )

    if tox_sections:
        res.warn(
            f"tox-related sections {sorted(tox_sections.keys())}; tox does not currently support pyproject.toml "
            f"except by way of `legacy_tox_in`; see https://tox.wiki/en/4.3.3/config.html#pyproject-toml",
            path=pth,
        )


def process_file(res: Result, pth: Path):
    if pth.name in CONFIG_FILES:
        process_config_file(res, pth)
        return
    raise NotImplementedError(f"Can't process {pth}")
