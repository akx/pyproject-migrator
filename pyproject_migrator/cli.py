import argparse
import logging
import sys
from pathlib import Path

import tomlkit

from pyproject_migrator.convert import CONFIG_FILES, process_file
from pyproject_migrator.result import Result


def parse_args():
    ap = argparse.ArgumentParser()
    ap.add_argument("src", nargs="+", help="Source files/directories")
    return ap.parse_args()


def cli() -> None:
    logging.basicConfig(level=logging.INFO, format="%(levelname)s: %(message)s")
    args = parse_args()
    res = Result()
    for src in args.src:
        pth = Path(src)
        if pth.is_dir():
            for config_file_name in sorted(CONFIG_FILES):
                config_file = pth / config_file_name
                if config_file.exists():
                    process_file(res, config_file)
        elif pth.is_file():
            process_file(res, pth)
    tomlkit.dump(res.dest, sys.stdout)
