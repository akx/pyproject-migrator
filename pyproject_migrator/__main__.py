from pyproject_migrator.cli import cli

if __name__ == "__main__":  # pragma: no cover
    cli()
