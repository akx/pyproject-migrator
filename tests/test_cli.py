import logging
from pathlib import Path
from pyproject_migrator.cli import cli

import tomlkit

victim_path = Path(__file__).parent / "victim"


def test_cli(monkeypatch, capsys, caplog, snapshot):
    caplog.set_level(logging.INFO)
    monkeypatch.setattr("sys.argv", ["tool", str(victim_path)])
    cli()
    # Snip out pathnames for snapshot comparison...
    messages = sorted(m.partition(": ")[-1] for m in caplog.messages)
    assert messages == snapshot(name="log")
    assert "setuptools sections ['metadata', 'options']" in caplog.text
    assert "flake8-related sections ['flake8']" in caplog.text
    parsed = tomlkit.loads(capsys.readouterr()[0])
    assert parsed == snapshot(name="result")
