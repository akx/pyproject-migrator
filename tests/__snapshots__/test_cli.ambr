# serializer version: 1
# name: test_cli[log]
  list([
    'Appending values in tool.pytest.ini_options->filterwarnings; make sure the order is correct',
    "flake8-related sections ['flake8']; porting depends on which toolset to use. If you're transitioning to ruff, see https://pypi.org/project/flake8-to-ruff/",
    "setuptools sections ['metadata', 'options']; porting depends on which toolset to use. For example, hatch (https://pypi.org/project/hatch/) can port metadata to pyproject.toml.",
  ])
# ---
# name: test_cli[result]
  TOMLDocument({
    'tool': Table({
      'isort': Table({
        'multi_line_output': 3,
        'profile': 'black',
      }),
      'mypy': Table({
        'check_untyped_defs': False,
        'implicit_reexport': False,
        'no_implicit_optional': True,
        'overrides': AoT([
          Table({
            'ignore_missing_imports': True,
            'module': 'colors',
          }),
        ]),
        'pretty': True,
        'python_version': '3.5',
        'show_column_numbers': True,
        'show_error_codes': True,
        'show_error_context': True,
        'strict_equality': True,
        'warn_no_return': True,
        'warn_redundant_casts': False,
        'warn_return_any': True,
        'warn_unreachable': True,
        'warn_unused_configs': True,
        'warn_unused_ignores': False,
      }),
      'pytest': Table({
        'ini_options': Table({
          'doctest_optionflags': Array([
            'NORMALIZE_WHITESPACE',
            'IGNORE_EXCEPTION_DETAIL',
            'ALLOW_UNICODE',
          ]),
          'env': Array([
            'AWS_ACCESS_KEY_ID=foobar_key',
            'AWS_SECRET_ACCESS_KEY=foobar_secret',
          ]),
          'filterwarnings': Array([
            'ignore:.*imp module is deprecated.*:DeprecationWarning',
            'error',
            'default:Exception ignored.*FileIO.*',
          ]),
          'norecursedirs': Array([
            'bower_components',
            'node_modules',
            'foo',
            'bar',
          ]),
        }),
      }),
    }),
  })
# ---
